// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// REQUIRES: aarch64-registered-target
// RUN: %clang_cc1 -triple aarch64-none-linux-gnu \
// RUN:   -target-feature +sme-i16i64 -fsyntax-only -verify %s

#include <arm_sme.h>

__attribute__((arm_streaming, arm_shared_za))
void test_svaddha_za32_s32(svbool_t pn, svbool_t pm, svint32_t zn) {
  // expected-error@+1 {{argument value 4 is outside the valid range [0, 3]}}
  svaddha_za32_s32_m(4, pn, pm, zn);
}

__attribute__((arm_streaming, arm_shared_za))
void test_svaddha_za32_u32(svbool_t pn, svbool_t pm, svuint32_t zn) {
  // expected-error@+1 {{argument value 4 is outside the valid range [0, 3]}}
  svaddha_za32_u32_m(4, pn, pm, zn);
}

__attribute__((arm_streaming, arm_shared_za))
void test_svaddha_za64_s64(svbool_t pn, svbool_t pm, svint64_t zn) {
  // expected-error@+1 {{argument value 8 is outside the valid range [0, 7]}}
  svaddha_za64_s64_m(8, pn, pm, zn);
}

__attribute__((arm_streaming, arm_shared_za))
void test_svaddha_za64_u64(svbool_t pn, svbool_t pm, svuint64_t zn) {
  // expected-error@+1 {{argument value 8 is outside the valid range [0, 7]}}
  svaddha_za64_u64_m(8, pn, pm, zn);
}

__attribute__((arm_streaming, arm_shared_za))
void test_svaddva_za32_s32(svbool_t pn, svbool_t pm, svint32_t zn) {
  // expected-error@+1 {{argument value 4 is outside the valid range [0, 3]}}
  svaddva_za32_s32_m(4, pn, pm, zn);
}

__attribute__((arm_streaming, arm_shared_za))
void test_svaddva_za32_u32(svbool_t pn, svbool_t pm, svuint32_t zn) {
  // expected-error@+1 {{argument value 4 is outside the valid range [0, 3]}}
  svaddva_za32_u32_m(4, pn, pm, zn);
}

__attribute__((arm_streaming, arm_shared_za))
void test_svaddva_za64_s64(svbool_t pn, svbool_t pm, svint64_t zn) {
  // expected-error@+1 {{argument value 8 is outside the valid range [0, 7]}}
  svaddva_za64_s64_m(8, pn, pm, zn);
}

__attribute__((arm_streaming, arm_shared_za))
void test_svaddva_za64_u64(svbool_t pn, svbool_t pm, svuint64_t zn) {
  // expected-error@+1 {{argument value 8 is outside the valid range [0, 7]}}
  svaddva_za64_u64_m(8, pn, pm, zn);
}
